### DISCUSSION TITLE
Is it possible to trigger a realtime event on database record updates?

### DISCUSSION THREAD

**Discussion by SkYNewZ** [pocketbase/pocketbase#4427] [Q&A] [answered]
I have processes running in the background that update elements in the database directly (via `*daos.Dao`). On the other hand, I have a React interface that subscribes to the changes corresponding to this collection.

However, it seems that realtime events are only triggered by Pocketbase API changes. Is it possible to force a new realtime event when a database change is made? Is it possible to do something like this?

Thanks

```go
app.OnModelAfterCreate(db.TasksCollectionName).Add(func(e *core.ModelEvent) error {
		return app.OnRecordAfterUpdateRequest(db.TasksCollectionName).Trigger(&core.RecordUpdateEvent{
			BaseCollectionEvent: core.BaseCollectionEvent{},
			HttpContext:         nil,
			Record:              nil,
			UploadedFiles:       nil,
		})
	})
```

**Comment by ganigeorgiev** [comment]
>  it seems that realtime events are only triggered by Pocketbase API changes

That is not true. The realtime events listen to the `OnModel*` hooks and they don't rely on the client API requests.

> I have processes running in the background that update elements in the database directly

Please show the code (_note that raw SQL queries doesn't fire realtime events_).

**Reply by ganigeorgiev** [reply]
Additionally, by _"processes running in the background"_ do you mean a second new process in addition to the one started with `./pocketbase serve` or just extra goroutine(s) within the same process?

If it is a new process, then Yes, realtime events (or any other app event hook) will not be fired and there is no builtin method to trigger app event hooks across different processes.

**Comment by SkYNewZ** [comment]
> Additionally, by "processes running in the background" do you mean a second process in addition to the one started with ./pocketbase serve?

My entire app is based on top of Pocketbase, not with Javascript hooks or default binary.

The methods `CreateOrUpdate` and `UpdateStatus` does not seems to trigger these events since I don't see any "stains" appearing. `*daos.Dao` is from the base `app.Dao()` method.

### snippet of UI
```typescript
const Tasks = ({ playlistId }: Props) => {
  const { pb } = usePocketBase();
  const [tasks, setTasks] = useState<Task[]>([]);
  const { t } = useTranslation("tasks");

  useEffect(() => {
    // get tasks
    pb.collection("tasks")
      .getList<Task>(1, 15, {
        sort: "-created",
        filter: `playlist = "${playlistId}"`,
      })
      .then((res) => setTasks(res.items));

    // Subscribe to tasks changes
    pb.collection("tasks").subscribe<Task>("*", (e) => {
      if (e.action === "create") {
        setTasks((prev) => [e.record, ...prev]); // Add new task to the top
      }
      if (e.action === "update") {
        setTasks((prev) =>
          prev.map((task) => (task.id === e.record.id ? e.record : task)),
        );
      }
      if (e.action === "delete") {
        setTasks((prev) => prev.filter((task) => task.id !== e.record.id));
      }
    });

    // Unsubscribe from tasks changes
    return () => {
      pb.collection("tasks").unsubscribe();
    };
  }, [pb, playlistId]);
```

### snippet of code updating my record
```go
// import…
package db

// TasksCollectionName is the name of the collection that stores tasks history data.
const TasksCollectionName string = "tasks"

var _ Model = (*Task)(nil)

// Task store history of scheduled tasks.
type Task struct {
	models.BaseModel
	record *models.Record // original record field

	TaskID     string        `db:"task_id"   json:"task_id"`
	TaskType   string        `db:"task_type" json:"task_type"`
	Status     types.JsonRaw `db:"status"    json:"status"`
	PlaylistID string        `db:"playlist"  json:"playlist"` // this is the playlist entity ID, not the YouTube playlist ID
	Payload    types.JsonRaw `db:"payload"   json:"payload"`
}

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	"github.com/pocketbase/pocketbase/tools/types"
)

var (
	_tasksRepositoryOnce = new(sync.Once)
	_tasksRepository     *tasksRepository
)

type tasksRepository struct {
	dao *daos.Dao
}

// NewTasksRepository creates a new tasks repository.
//
//go:generate go run github.com/vburenin/ifacemaker --file $GOFILE --struct tasksRepository --iface TasksRepository --pkg db --output tasks_repository_iface.go
func NewTasksRepository(dao *daos.Dao) TasksRepository {
	_tasksRepositoryOnce.Do(func() {
		_tasksRepository = &tasksRepository{dao: dao}
	})

	return _tasksRepository
}

// GetByTaskID returns a task record by its task_id.
func (r *tasksRepository) GetByTaskID(taskID string) (*Task, error) {
	record, err := r.dao.FindFirstRecordByData(TasksCollectionName, "task_id", taskID)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, ErrTaskNotFound
	}

	if err != nil {
		return nil, fmt.Errorf("could not find task: %w", err)
	}

	return newTaskFromRecord(record), nil
}

// CreateOrUpdate creates or updates a task record.
func (r *tasksRepository) CreateOrUpdate(task *Task) error {
	created, err := r.GetByTaskID(task.TaskID)

	// task not found, create it
	if errors.Is(err, ErrTaskNotFound) {
		if err := r.dao.Save(task); err != nil {
			return fmt.Errorf("failed to create task: %w", err)
		}

		return nil
	}

	// task found, update it
	created = &Task{
		BaseModel:  created.BaseModel,
		TaskID:     task.TaskID,
		TaskType:   task.TaskType,
		Status:     task.Status,
		PlaylistID: task.PlaylistID,
		Payload:    task.Payload,
	}
	if err := r.dao.Save(created); err != nil {
		return fmt.Errorf("failed to update task: %w", err)
	}

	return nil
}

// UpdateStatus updates the status of a task record.
func (r *tasksRepository) UpdateStatus(taskID string, status []byte) error {
	record, err := r.GetByTaskID(taskID)
	if err != nil {
		return err
	}

	record.Status = status
	if err := r.dao.Save(record); err != nil {
		return fmt.Errorf("could not save task: %w", err)
	}

	return nil
}

// DeleteWhenOlderThan deletes tasks older than the given age.
// If taskType is provided, only tasks of that type will be deleted.
func (r *tasksRepository) DeleteWhenOlderThan(ctx context.Context, age time.Duration, taskType ...string) (int64, error) {
	d, err := types.ParseDateTime(time.Now().Add(-age).UTC()) // Pocketbase stores time in UTC
	if err != nil {
		return 0, fmt.Errorf("could not parse date: %w", err)
	}

	expr := []dbx.Expression{dbx.NewExp("created<{:created}", dbx.Params{"created": d})}
	if len(taskType) > 0 {
		expr = append(expr, dbx.In("task_type", interfaceValues(taskType)...))
	}

	res, err := r.dao.DB().Delete(TasksCollectionName, dbx.And(expr...)).WithContext(ctx).Execute()
	if err != nil {
		return 0, fmt.Errorf("could not delete tasks: %w", err)
	}

	rows, err := res.RowsAffected()
	if err != nil {
		rows = -1 // unknown
	}

	return rows, nil
}

func interfaceValues[T any](values []T) []any {
	vv := make([]interface{}, len(values))
	for i, v := range values {
		vv[i] = v
	}
	return vv
}

```

**Reply by ganigeorgiev** [reply]
~The realtime events doesn't fire because you are using custom struct models.
As mentioned [in the docs](https://pocketbase.io/docs/go-custom-models/#working-with-the-custom-model-struct), records related hooks fire only for `models.Record`.~

See the edit below (I forgot that we had implemented an exception for the realtime events).

**Reply by ganigeorgiev** [reply]
Additionally, note that raw DB queries like the `DB().Delete(...)` will not trigger the app hooks.
The create/update/delete model hooks are fired only when using `Dao.Save` or `Dao.Delete` methods.

**Reply by SkYNewZ** [reply]
> Additionally, note that raw DB queries like the `DB().Delete(...)` will not trigger the app hooks.

Thanks for the tip. Deletion in this case does not need to be displayed in real time (some garbage collector).

> The realtime events doesn't fire because you are using custom struct models.

Okay my bad… So something like this should trigger realtime events? But I was already using `Dao.Save` before?

```diff
diff --git i/internal/db/tasks_repository.go w/internal/db/tasks_repository.go
index 3de5c73..a71a8cb 100644
--- i/internal/db/tasks_repository.go
+++ w/internal/db/tasks_repository.go
@@ -10,6 +10,7 @@ import (
 
 	"github.com/pocketbase/dbx"
 	"github.com/pocketbase/pocketbase/daos"
+	"github.com/pocketbase/pocketbase/models"
 	"github.com/pocketbase/pocketbase/tools/types"
 )
 
@@ -47,30 +48,18 @@ func (r *tasksRepository) GetByTaskID(taskID string) (*Task, error) {
 	return newTaskFromRecord(record), nil
 }
 
-// CreateOrUpdate creates or updates a task record.
-func (r *tasksRepository) CreateOrUpdate(task *Task) error {
-	created, err := r.GetByTaskID(task.TaskID)
-
-	// task not found, create it
-	if errors.Is(err, ErrTaskNotFound) {
-		if err := r.dao.Save(task); err != nil {
-			return fmt.Errorf("failed to create task: %w", err)
-		}
-
-		return nil
-	}
-
-	// task found, update it
-	created = &Task{
-		BaseModel:  created.BaseModel,
-		TaskID:     task.TaskID,
-		TaskType:   task.TaskType,
-		Status:     task.Status,
-		PlaylistID: task.PlaylistID,
-		Payload:    task.Payload,
-	}
-	if err := r.dao.Save(created); err != nil {
-		return fmt.Errorf("failed to update task: %w", err)
+// Create creates a new task record.
+func (r *tasksRepository) Create(task *Task) error {
+	collection, _ := r.dao.FindCollectionByNameOrId(TasksCollectionName)
+	record := models.NewRecord(collection)
+	record.Set("task_id", task.TaskID)
+	record.Set("task_type", task.TaskType)
+	record.Set("status", task.Status)
+	record.Set("playlist", task.PlaylistID)
+	record.Set("payload", task.Payload)
+
+	if err := r.dao.SaveRecord(record); err != nil {
+		return fmt.Errorf("failed to create task: %w", err)
 	}
 
 	return nil
@@ -78,13 +67,14 @@ func (r *tasksRepository) CreateOrUpdate(task *Task) error {
 
 // UpdateStatus updates the status of a task record.
 func (r *tasksRepository) UpdateStatus(taskID string, status []byte) error {
-	record, err := r.GetByTaskID(taskID)
+	task, err := r.GetByTaskID(taskID)
 	if err != nil {
 		return err
 	}
 
-	record.Status = status
-	if err := r.dao.Save(record); err != nil {
+	record := task.Record()
+	record.Set("status", status)
+	if err := r.dao.SaveRecord(record); err != nil {
 		return fmt.Errorf("could not save task: %w", err)
 	}
 

```

**Reply by ganigeorgiev** [reply, marked as answer]
This thread become a little messy. Sorry for the confusion earlier.

I forgot that we had implemented in the past a special case for the realtime events to support custom model structs:
https://github.com/pocketbase/pocketbase/blob/27b6d0c505efbf7aee3cd545e104976137c9743b/apis/realtime.go#L339-L350

To summarize - realtime events should work when using your custom model calling `r.dao.Save(yourTask)`.

If it doesn't work then double check that the subscribed client satisfy the List (for the `*` topic) and View (when using a concrete record id) API rules of the `tasks` collection and that your frontend code works correctly (eg. inspect the realtime connect/disconnect events printed in the terminal during dev mode).

**Reply by ganigeorgiev** [reply]
If you are still not able to resolve it, please create a **minimal** reproducible repo and I'll try to investigate it in more details tomorrow.

**Reply by SkYNewZ** [reply]
Okay. I'm investigating. For other collection, realtime works great. Also for those "working" collections, user creates the record via **authenticated** API request.

I still struggling to know why it can't work on these specific record created on the back side. Here is a snippet of database schema. **`List` and `View` rules are well configured because as an authenticated user I can list `tasks`**, I just need to refresh the page after doing something that create a `task`. In my tests I'm the only one user.

If I can't figure it out I will create a minimal repo. Thanks for your help anyway, as always.

<img width="679" alt="image" src="https://github.com/pocketbase/pocketbase/assets/26750012/311e5e43-6d9a-4f07-a8c6-8fdc77ccfe6d">
<img width="664" alt="image" src="https://github.com/pocketbase/pocketbase/assets/26750012/79333cf5-ce3e-4643-b8e2-8a01377dbec0">
<img width="663" alt="image" src="https://github.com/pocketbase/pocketbase/assets/26750012/4fc5b9c9-5b9c-4cff-8962-d3bb0d5ce07e">
<img width="668" alt="image" src="https://github.com/pocketbase/pocketbase/assets/26750012/d2f63e14-2acd-417e-b010-aa3c70d986f5">

**Reply by SkYNewZ** [reply]
By using breakpoints on https://github.com/pocketbase/pocketbase/blob/27b6d0c505efbf7aee3cd545e104976137c9743b/apis/realtime.go#L258 and do trigger a `task` (custom model) creation on the back side, the breakpoint is never handled.

**Reply by ganigeorgiev** [reply]
This is only when an auth record changes in order to update the subscribed client auth state.

There is another model hook below this one that takes care for broadcasting the record changes. For the update operation that would be:
https://github.com/pocketbase/pocketbase/blob/27b6d0c505efbf7aee3cd545e104976137c9743b/apis/realtime.go#L296

**Reply by SkYNewZ** [reply]
I think I've got an idea. In my final binary I expose several commands that I've added to Pocketbase's `RootCmd`.

- `serve` (the default serve command)
- `worker` (a "worker" that listen for some things and do some async work)

The `worker` :
- receives the task to be processed
- saves the task in the database via Dao.

But these 2 instances of `Pocketbase` run in parallel (`./binary serve` and `./binary worker`) based on the same `--dir` flag. It's logical that the events of one don't trigger the events of the other... sorry for the noise.

**Reply by SkYNewZ** [reply]
Each subcommand has his own Pocketbase instance :/
```
PocketBase CLI

Usage:
  hop [command]

Available Commands:
  admin       Manages admin accounts
  migrate     Executes app DB migration scripts
  scheduler   Start scheduler and scheduler config
  serve       Starts the web server (default to 127.0.0.1:8090 if no domain is specified)
  worker      Start tasks queue worker

Flags:
      --dev                    enable dev mode, aka. printing logs and sql statements to the console (default true)
      --dir string             the PocketBase data directory (default "/Users/quentin/src/skynewz/hop/pb_data")
      --encryptionEnv string   the env variable whose value of 32 characters will be used 
                               as encryption key for the app settings (default none)
  -h, --help                   help for youtube-plus
      --redis-addr string      address of redis server to connect to (default "127.0.0.1:6379")
  -v, --version                version for hop

Use "hop [command] --help" for more information about a command.

```

**Reply by ganigeorgiev** [reply]
Yes, if you have 2 different processes it will not work since the app hooks are managed on application level.

How the `worker` command is invoked and how it receives the tasks to process?
Is it not possible to start it as part of the same process in a separate goroutine?

**Reply by SkYNewZ** [reply]
> How the `worker` command is invoked

All my app stack (`Pocketbase "core"`, `worker`, `scheduler`) is launched via the same base Docker image but different argument. All with the same `--dir` flag.

> How it receives the tasks to process?

The worker is based on https://github.com/hibiken/asynq. It receives tasks from
	- authenticated user requests to Pocketbase custom routes for which I create a job
	- a scheduler (which is also a subcommand added to the Pocketbase's `RootCmd`) which reads job config from a dedicated Pocketbase collection, creates and schedules job.

A `task` like I described before is in fact a job result stored in another dedicated collection used to be exposed to the user who triggered the "asynq" job. I wanted to ensure that when the user triggered a job by clicking on a button, the result of this job was instantly displayed in the list of 'tasks' in progress in relation to the desired action. Unfortunately, in my architecture, this doesn't seem possible.

> Is it not possible to start it as part of the same process in a separate goroutine?

It would be possible to run the `worker` in the same instance of Pocketbase, but I wanted it to be completely independent.
- Pocketbase exposes the API and the frontend
- the worker can be started multiple times to share the workload. 
- the scheduler is also completely independent and periodically reads the Pocketbase database.
I just wanted each component to be standalone. I can have several `workers` without having "several" `frontents`, for example.

**Reply by SkYNewZ** [reply]
The `worker` could create or update the record using an http request to Pocketbase API with a static admin token but I don't think that makes sense.

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what do you mean by "share the workload" but having multiple long running processes doesn't necessary makes it more efficient, and specifically because PocketBase can't scale on more than 1 server, it actually may ended up consuming more resources because each command will have to bootstrap their own `core.App`, store, connections pool, etc.

> I just wanted each component to be standalone.

They can still be standalone and in their own package(s) but instead of being executed separately by the Docker container, you can simply initialize them from the same process, eg.

```go
func main() {
	app := pocketbase.New()

	go func() {
		worker.Start(app)
	}()

	go func() {
		scheduler.Start(app)
	}()

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by SkYNewZ** [reply]
I wanted to avoid monolithic application. By sharing the workload I mean all jobs can be shared across multiple worker instances. But you are right each app has his own bootstrap etc.

I was afraid that if a component panics it will break the entire application instead of just the exact component (which I could just restart). But maybe I have over-engineered my app trying a micro-service architecture.

I also wanted each component has his own flag set (redis address, dry run toggle, etc). You makes me rethinking all the app 😅 is it worth to split components that all relies on the same database provided by Pocketbase…

**Reply by SkYNewZ** [reply]
I'm trying to start `scheduler` and `worker` on the same Pocketbase instance. On the very beginning of Pocketbase declaration I used to call custom flag registration like so:

<details>
<summary>Snippet</summary>

```go
func RegisterFlags(app *pocketbase.PocketBase) {
	app.RootCmd.PersistentFlags().String(RedisAddrFlag, getEnvDefaultString("REDIS_ADDR", "127.0.0.1:6379"), "address of redis server to connect to")
	app.RootCmd.PersistentFlags().Duration(SyncIntervalFlag, 10*time.Minute, "Scheduler sync interval")
	app.RootCmd.PersistentFlags().Int(NumWorkersFlag, 3, "Number of workers")
	app.RootCmd.PersistentFlags().Bool(DryRunFlag, false, "Dry run mode. No real action will be taken on YouTube.")
}
```

</details>

This used to work fine because `worker` and `scheduler` were sub-commands of the main binary. How can I access these flags? On `OnAfterBootstrap` the flags are not set. Should I use `OnBeforeServe`?

**Reply by ganigeorgiev** [reply]
You can register them right after `pocketbase.New()` but you'll have to call `app.RootCmd.ParseFlags` manually (see the example [main.go](https://github.com/pocketbase/pocketbase/blob/master/examples/base/main.go)).

**Reply by SkYNewZ** [reply]
I've managed to refactor my application to use a single binary and benefit from all the events available. Thanks for your help.
